### ISSUE TITLE
Add Keep-Alive configurable setting for terminal

### ISSUE THREAD

**Issue by zyzhu2000** [jupyter/notebook#1072]
I run notebook behind nginx.  One problem I have with the terminal function is that nginx will close the websocket connection after 60 seconds. It's clearly not a good solution to make the timeout longer in nginx. Another solution is for the backend periodically send a websocket ping to the web browser when there is no activity.  

For this, I just added some code to terminado to support a 'keep_alive_time.' See details here:
https://github.com/takluyver/terminado/commit/1c591faeac3efafe70e34d8b79d90c32d747ad53

Basically if keep_alive_time=x, it means it will send a websocket ping if there is no activity for x seconds. If x is 0, then the ping business is disabled.

What I want is for notebook to expose this setting in the config file, so that people who need them can turn it on.

**Comment by takluyver** [comment]
Seems reasonable. Do you want to make a PR? It will need to be passed through a few layers in `notebookapp.py` - I'm happy to explain that a bit if you want.

We'll need to do a new release of terminado before merging this.

**Comment by zyzhu2000** [comment]
I will try when I get a moment. Thanks

Sent from Outlook Mobile




On Tue, Feb 9, 2016 at 9:47 AM -0800, "Thomas Kluyver" <notifications@github.com> wrote:





Seems reasonable. Do you want to make a PR? It will need to be passed through a few layers in `notebookapp.py` - I'm happy to explain that a bit if you want.

We'll need to do a new release of terminado before merging this.

---

**Comment by takluyver** [comment]
If you're seeing that with the terminal websockets, do you see a similar thing in the notebook if you leave an open notebook inactive for 60 seconds? The kernel messages also go over a websocket. But maybe the auto-reconnect code we have there is sufficient to reopen the socket.

I wonder if we should send a PR upstream to tornado for a similar option to ping any websocket that is open and inactive for > X seconds.

@minrk probably has a better idea than me whether that's reasonable.

**Comment by zyzhu2000** [comment]
No I don't see this with the notebook. It is just the terminal.

Sent from Outlook Mobile

    _____________________________

**Comment by minrk** [comment]
The kernel websockets already are sending ws pings, for precisely this reason. This is implemented in [WebSocketMixin](https://github.com/jupyter/notebook/blob/4.1.0/notebook/base/zmqhandlers.py#L99). It seems like the terminal websockets should already be inheriting this behavior, though.

**Comment by takluyver** [comment]
`TermSocket.open` doesn't call its super method, which is what would set up the pinging.

**Comment by takluyver** [comment]
@zyzhu2000 I might back out your changes in terminado and just fix it so that it can use the pinging already implemented in Jupyter. Sorry, I didn't realise that Jupyter had already got this code.

**Comment by takluyver** [comment]
With takluyver/terminado#28, I think this should work with no changes to Jupyter.

**Comment by takluyver** [comment]
@minrk now I look at this some more, I don't understand why Jupyter's pinging is not already working. `WebSocketMixin` is before `TermSocket` in the base classes, so I think I was wrong about the lack of a super call preventing pinging from being set up. Here's the MRO:

```
In [2]: from notebook.terminal.handlers import TermSocket

In [5]: type.mro(TermSocket)
Out[5]: 
[notebook.terminal.handlers.TermSocket,
 notebook.base.zmqhandlers.WebSocketMixin,
 notebook.base.handlers.IPythonHandler,
 notebook.base.handlers.AuthenticatedHandler,
 terminado.websocket.TermSocket,
 tornado.websocket.WebSocketHandler,
 tornado.web.RequestHandler,
 object]
```

So I'm not sure why the pinging doesn't already work for terminal websockets.

**Comment by JamiesHQ** [comment]
We're doing a little housekeeping on our issue log and noticed this thread from 2016. Has this issue been resolved to your satisfaction and can it be closed? thanks!

**Comment by cailiang9** [comment]
I'm using chrome browser with a proxy to open the jupyter terminal.
When I switch from terminal tab to another, the websocket becomes closed. To interact, the only thing I can do is to refresh. And many information in the buffer is lost.
Are there any configurations to avoid this?

**Comment by akarukappadath** [comment]
I'm also experiencing the same issue, but I'm using apache instead of nginx. Has anyone been working on this?

**Comment by PMende** [comment]
I also experience this issue. We're running single-user jupyter notebook servers behind a traefik load balancer inside EKS.
